src/sgame/sg_bot_ai.cpp

```diff
@@ -577,11 +577,17 @@ AINodeStatus_t BotActionActivateUpgrade( gentity_t *self, AIGenericNode_t *node
 	AIActionNode_t *action = ( AIActionNode_t * ) node;
 	upgrade_t u = ( upgrade_t ) AIUnBoxInt( action->params[ 0 ] );
 
-	if ( !BG_UpgradeIsActive( u, self->client->ps.stats ) &&
-		BG_InventoryContainsUpgrade( u, self->client->ps.stats ) )
+	if ( !BG_InventoryContainsUpgrade( u, self->client->ps.stats ) )
 	{
-		BG_ActivateUpgrade( u, self->client->ps.stats );
+		return STATUS_FAILURE;
+	}
+
+	if ( BG_UpgradeIsActive( u, self->client->ps.stats ) )
```

Comment:
I suppose the only way this could happen is if you tried to use medkit/grenade/firebomb more than once in a single frame. (This is not a criticism, just an observation.)

Medkits can fail to activate if you have full health, but the behavior tree wouldn't detect this.

Reply:
Honestly, I don't know enough that part of the code for now, so I'm trying to get the problem fixed once and for once. The symptoms really don't make any sense when the BT shows a bug, and I would rather have a pedantic API that makes it obvious what should happen if an action explicitly failed that to continue with that Schrödinger cat API...

There are, AFAIK, 4 upgrades:

1. grenade
2. firebomb
3. medkit
4. jetpack

Of those, grenade and firebomb have specific code, but do not really make sense as upgrades anyway. IMO they should be put in one's hands, and then thrown, possibly adding some better time and direction control, which would possibly be more fitting as weapon (don't know, really). Those are the ones which actually triggered the bug.
You covered medkit's case.
Jetpack is not yet usable by bots, but this could happen. When it becomes, I don't think it will belong to BT anyway, but rather to pathfinding, and is the only case in which one would want to enable an already enabled upgrade.

Reply:
Was going to edit, but changes in between, so separate post:

You covered medkit's case, but it is worth mentioning that it could use the same BT code reduction as the grenades received, that is (not tested, might trigger other problems):

```
decorator return( STATUS_FAILURE )
{
	condition ( percentHealth( E_SELF ) <= 0.5  )
	{
		action activateUpgrade( UP_MEDKIT )
	}
}
```

Instead of:

```
decorator return( STATUS_FAILURE )
{
	condition ( percentHealth( E_SELF ) <= 0.5 && haveUpgrade( UP_MEDKIT ) )
	{
		action activateUpgrade( UP_MEDKIT )
	}
}
```